src/relations/peers.py

```diff
@@ -0,0 +1,255 @@
+        - relation-joined
+        - relation-changed
+    """
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, PEER_RELATION_NAME)
+
+        self.charm = charm
+
+        self.framework.observe(charm.on[PEER_RELATION_NAME].relation_created, self._on_created)
+        self.framework.observe(charm.on[PEER_RELATION_NAME].relation_joined, self._on_changed)
+        self.framework.observe(charm.on[PEER_RELATION_NAME].relation_changed, self._on_changed)
+
+    @property
+    def app_databag(self):
+        """Returns the app databag for the Peer relation."""
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        if peer_relation is None:
+            return {}
+        return peer_relation.data[self.charm.app]
+
+    @property
+    def unit_databag(self):
+        """Returns the app databag for the Peer relation."""
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        if peer_relation is None:
+            return {}
+        return peer_relation.data[self.charm.unit]
+
+    def _on_created(self, event: RelationCreatedEvent):
+        if not self.charm.unit.is_leader():
+            return
+
+        try:
+            cfg = self.charm.read_pgb_config()
+        except FileNotFoundError:
+            # If there's no config, the charm start hook hasn't fired yet, so defer until it's
+            # available.
+            event.defer()
+            return
+
+        self.update_cfg(cfg)
+
+        try:
+            if self.charm.backend:
+                # update userlist only if backend relation exists.
+                self.update_auth_file(self.charm.read_auth_file())
+        except FileNotFoundError:
+            # Auth files will be added to the databag once they're available
+            pass
+
+    def _on_changed(self, event: RelationChangedEvent):
+        """If the current unit is a follower, write updated config and auth files to filesystem."""
+        if self.charm.unit.is_leader():
+            try:
+                cfg = self.charm.read_pgb_config()
+            except FileNotFoundError:
+                # If there's no config, the charm start hook hasn't fired yet, so defer until it's
+                # available.
+                event.defer()
+                return
+
+            self.update_cfg(cfg)
+            return
+
+        if cfg := self.app_databag.get(CFG_FILE_DATABAG_KEY):
+            self.charm.render_pgb_config(PgbConfig(cfg))
+
+        if auth_file := self.app_databag.get(AUTH_FILE_DATABAG_KEY):
+            self.charm.render_auth_file(auth_file)
+
+        if cfg is not None or auth_file is not None:
+            try:
+                # returns an error if this is fired before on_pebble_ready.
+                self.charm.reload_pgbouncer()
+            except ConnectionError:
+                # TODO find a better error to use
+                event.defer()
+
+    def set_secret(self, scope:str, key:str, value:str):
```

Comment:
Wouldn't be better to move this to `charm.py` or another place, as it should be changed to use Juju secrets instead of relation data in the future?

Reply:
I think in future you're right, but I want to keep updates for this databag mostly in this file. Would a wrapper method in charm.py work?

Reply:
No worries. It can be done in the future.